In [1]:
!pip install keras


Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.0.1 -> 25.2
[notice] To update, run: C:\Program Files\Python313\python.exe -m pip install --upgrade pip


In [2]:
!pip install tensorflow

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.0.1 -> 25.2
[notice] To update, run: C:\Program Files\Python313\python.exe -m pip install --upgrade pip


In [138]:
#importing Libraries

In [58]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import accuracy_score,precision_score,recall_score,confusion_matrix,f1_score
from sklearn.model_selection import train_test_split,GridSearchCV
from sklearn.preprocessing import LabelEncoder,StandardScaler
from keras.models import Sequential
from keras.layers import Dense,Dropout
from tensorflow.keras.optimizers import Adam
import warnings
warnings.filterwarnings("ignore")

In [139]:
#Loading the data frame

In [2]:
df=pd.read_csv("Alphabets_data.csv")

In [3]:
df.shape

(20000, 17)

In [4]:
df.head()

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10

In [5]:
df.tail()

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  \
19995      D     2     2      3       3      2     7     7      7      6   
19996      C     7    10      8       8      4     4     8      6      9   
19997      T     6     9      6       7      5     6    11      3      7   
19998      S     2     3      4       2      1     8     7      2      6   
19999      A     4     9      6       6      2     9     5      3      1   

       xybar  x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
19995      6       6       4      2       8      3       7  
19996     12       9      13      2       9      3       7  
19997     11       9       5      2      12      2       4  
19998     10       6       8      1       9      5       8  
19999      8       1       8      2       7      2       8

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   letter  20000 non-null  object
 1   xbox    20000 non-null  int64 
 2   ybox    20000 non-null  int64 
 3   width   20000 non-null  int64 
 4   height  20000 non-null  int64 
 5   onpix   20000 non-null  int64 
 6   xbar    20000 non-null  int64 
 7   ybar    20000 non-null  int64 
 8   x2bar   20000 non-null  int64 
 9   y2bar   20000 non-null  int64 
 10  xybar   20000 non-null  int64 
 11  x2ybar  20000 non-null  int64 
 12  xy2bar  20000 non-null  int64 
 13  xedge   20000 non-null  int64 
 14  xedgey  20000 non-null  int64 
 15  yedge   20000 non-null  int64 
 16  yedgex  20000 non-null  int64 
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


In [140]:
# Checking for dummy variable

In [7]:
df.columns.duplicated().sum()


np.int64(0)

In [141]:
# Checking null values

In [8]:
df.isnull().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

In [9]:
df['letter'].unique()

array(['T', 'I', 'D', 'N', 'G', 'S', 'B', 'A', 'J', 'M', 'X', 'O', 'R',
       'F', 'C', 'H', 'W', 'L', 'P', 'E', 'V', 'Y', 'Q', 'U', 'K', 'Z'],
      dtype=object)

In [142]:
# Encoding the target variable using label encoder

In [10]:
LE=LabelEncoder()

In [11]:
df['letter']=LE.fit_transform(df['letter'])

In [12]:
df.dtypes

letter    int64
xbox      int64
ybox      int64
width     int64
height    int64
onpix     int64
xbar      int64
ybar      int64
x2bar     int64
y2bar     int64
xybar     int64
x2ybar    int64
xy2bar    int64
xedge     int64
xedgey    int64
yedge     int64
yedgex    int64
dtype: object

In [13]:
X=df.iloc[:,1:]
X

xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0         2     8      3       5      1     8    13      0      6      6   
1         5    12      3       7      2    10     5      5      4     13   
2         4    11      6       8      6    10     6      2      6     10   
3         7    11      6       6      3     5     9      4      6      4   
4         2     1      3       1      1     8     6      6      6      6   
...     ...   ...    ...     ...    ...   ...   ...    ...    ...    ...   
19995     2     2      3       3      2     7     7      7      6      6   
19996     7    10      8       8      4     4     8      6      9     12   
19997     6     9      6       7      5     6    11      3      7     11   
19998     2     3      4       2      1     8     7      2      6     10   
19999     4     9      6       6      2     9     5      3      1      8   

       x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0          10       8      0       8      0       8  
1           3       9      2       8      4      10  
2           3       7      3       7      3       9  
3           4      10      6      10      2       8  
4           5       9      1       7      5      10  
...       ...     ...    ...     ...    ...     ...  
19995       6       4      2       8      3       7  
19996       9      13      2       9      3       7  
19997       9       5      2      12      2       4  
19998       6       8      1       9      5       8  
19999       1       8      2       7      2       8  

[20000 rows x 16 columns]

In [143]:
# standardising the X data as mean=0 and std=1

In [14]:
SS=StandardScaler()

In [15]:
scaled_x=SS.fit_transform(X)

In [16]:
X=pd.DataFrame(scaled_x,columns=X.columns)
X

xbox      ybox     width    height     onpix      xbar      ybar  \
0     -1.057698  0.291877 -1.053277 -0.164704 -1.144013  0.544130  2.365097   
1      0.510385  1.502358 -1.053277  0.719730 -0.687476  1.531305 -1.075326   
2     -0.012309  1.199738  0.435910  1.161947  1.138672  1.531305 -0.645273   
3      1.555774  1.199738  0.435910  0.277513 -0.230939 -0.936631  0.644886   
4     -1.057698 -1.826464 -1.053277 -1.933571 -1.144013  0.544130 -0.645273   
...         ...       ...       ...       ...       ...       ...       ...   
19995 -1.057698 -1.523844 -1.053277 -1.049137 -0.687476  0.050543 -0.215220   
19996  1.555774  0.897117  1.428701  1.161947  0.225598 -1.430218  0.214833   
19997  1.033079  0.594497  0.435910  0.719730  0.682135 -0.443044  1.504991   
19998 -1.057698 -1.221224 -0.556881 -1.491354 -1.144013  0.544130 -0.215220   
19999 -0.012309  0.594497  0.435910  0.277513 -0.687476  1.037718 -1.075326   

          x2bar     y2bar     xybar    x2ybar    xy2bar     xedge    xedgey  \
0     -1.714360  0.344994 -0.917071  1.347774  0.034125 -1.305948 -0.219082   
1      0.137561 -0.495072  1.895968 -1.312807  0.514764 -0.448492 -0.219082   
2     -0.973591  0.344994  0.690380 -1.312807 -0.446513 -0.019764 -0.865626   
3     -0.232823  0.344994 -1.720796 -0.932724  0.995402  1.266419  1.074008   
4      0.507945  0.344994 -0.917071 -0.552641  0.514764 -0.877220 -0.865626   
...         ...       ...       ...       ...       ...       ...       ...   
19995  0.878329  0.344994 -0.917071 -0.172558 -1.888428 -0.448492 -0.219082   
19996  0.507945  1.605094  1.494105  0.967691  2.437316 -0.448492  0.427463   
19997 -0.603207  0.765028  1.092242  0.967691 -1.407789 -0.448492  2.367097   
19998 -0.973591  0.344994  0.690380 -0.172558  0.034125 -0.877220  0.427463   
19999 -0.603207 -1.755172 -0.113345 -2.072973  0.034125 -0.448492 -0.865626   

          yedge    yedgex  
0     -1.438153  0.122911  
1      0.120081  1.359441  
2     -0.269477  0.741176  
3     -0.659036  0.122911  
4      0.509640  1.359441  
...         ...       ...  
19995 -0.269477 -0.495354  
19996 -0.269477 -0.495354  
19997 -0.659036 -2.350149  
19998  0.509640  0.122911  
19999 -0.659036  0.122911  

[20000 rows x 16 columns]

In [17]:
y=df.iloc[:,:1]
y

letter
0          19
1           8
2           3
3          13
4           6
...       ...
19995       3
19996       2
19997      19
19998      18
19999       0

[20000 rows x 1 columns]

In [115]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42, stratify=y)

In [116]:
X_train

xbox      ybox     width    height     onpix      xbar      ybar  \
17133  0.510385  0.594497  1.428701  0.277513  1.138672  1.037718  0.214833   
17646  0.510385  1.199738 -0.060486  1.161947  0.682135 -1.923805 -0.215220   
18721  1.033079 -0.010743 -0.060486  2.046381  0.225598  0.544130 -1.075326   
4584  -1.580393 -1.221224 -1.549672 -1.933571 -1.144013 -0.443044  1.935044   
1744   1.555774  0.897117  2.421491  1.161947  1.138672 -1.430218  0.644886   
...         ...       ...       ...       ...       ...       ...       ...   
12162 -1.057698 -0.918603 -1.053277 -1.491354 -0.687476 -0.443044 -0.215220   
1591  -0.535004 -0.918603 -0.060486  0.277513 -0.230939 -0.443044  0.214833   
13697 -0.535004 -0.918603 -0.060486 -1.049137 -0.230939  1.037718  1.504991   
11686 -1.057698 -0.010743 -1.053277 -0.164704 -0.687476  0.050543 -0.215220   
18454  1.033079  0.594497  0.932305 -0.606921 -0.230939  2.518479 -2.365484   

          x2bar     y2bar     xybar    x2ybar    xy2bar     xedge    xedgey  \
17133 -0.232823  0.344994  1.092242 -0.552641 -1.407789  0.408963 -0.865626   
17646  0.878329 -0.915106 -0.515208 -0.172558  1.476040 -0.019764 -0.219082   
18721  0.137561 -0.075039  1.092242 -0.172558 -0.446513 -0.019764  0.427463   
4584  -0.973591 -0.915106 -0.113345  1.727857  0.034125 -0.448492  1.720552   
1744  -0.973591  0.765028  0.690380  0.967691  0.995402  0.837691 -1.512171   
...         ...       ...       ...       ...       ...       ...       ...   
12162  0.137561  0.344994  0.288517  0.207525  0.995402 -0.877220 -0.219082   
1591  -0.232823 -1.755172 -0.515208  0.587608  0.034125  2.123874  1.074008   
13697 -0.603207 -1.335139 -1.318933  0.967691 -0.446513  1.266419  1.074008   
11686 -1.714360  0.765028  1.895968 -0.172558  0.034125 -1.305948 -0.219082   
18454 -0.603207 -1.335139  0.690380 -1.312807  0.514764  1.266419 -2.805260   

          yedge    yedgex  
17133  0.120081  0.741176  
17646 -0.269477  1.977706  
18721  2.067874  0.122911  
4584  -1.438153  0.122911  
1744   0.120081 -1.731884  
...         ...       ...  
12162  0.120081  0.741176  
1591  -1.438153  0.122911  
13697 -1.438153  0.122911  
11686 -1.048594  0.122911  
18454 -1.048594  0.741176  

[14000 rows x 16 columns]

In [117]:
X.shape

(20000, 16)

In [118]:
y.shape

(20000, 1)

In [119]:
len(y['letter'].unique())

26

In [120]:
#  Build a Simple ANN Model
model = Sequential([
    Dense(128, activation='relu', input_shape=(X.shape[1],)),
    Dense(64, activation='relu'),
    Dense(len(y['letter'].unique()), activation='softmax')
])

In [121]:

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
# adam learning rate is 0.001 default

In [122]:

history = model.fit(X_train, y_train, validation_split=0.2, epochs=30, batch_size=32)


Epoch 1/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.5691 - loss: 1.6345 - val_accuracy: 0.7300 - val_loss: 0.9372
Epoch 2/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.7766 - loss: 0.7701 - val_accuracy: 0.8079 - val_loss: 0.6725
Epoch 3/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8271 - loss: 0.5892 - val_accuracy: 0.8400 - val_loss: 0.5395
Epoch 4/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.8558 - loss: 0.4825 - val_accuracy: 0.8657 - val_loss: 0.4546
Epoch 5/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8791 - loss: 0.4102 - val_accuracy: 0.8811 - val_loss: 0.4011
Epoch 6/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.8948 - loss: 0.3571 - val_accuracy: 0.8854 - val_loss: 0.3766
Epoch 7/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9065 - loss: 0.3166 - val_accuracy: 0.9011 - val_loss: 0.3351
Epoch 8/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.9149 - loss: 0.2819 - val_accuracy: 0.

In [123]:
history

In [124]:
scores=model.evaluate(X_train,y_train)

438/438 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9774 - loss: 0.0789


In [125]:
#print("%s:%.2f%%"%(model.metrics_names[1],scores[1]*100))

In [126]:
# Step 6: Evaluate the Model
loss, acc = model.evaluate(X_test, y_test, verbose=0)
print("Test Accuracy:", acc*100)

Test Accuracy: 94.1166639328003


In [127]:
y_pred=model.predict(X_test)
y_pred = np.argmax(y_pred, axis=1)

188/188 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step


In [128]:
y_pred

array([ 8,  2, 10, ..., 20,  1, 20], shape=(6000,))

In [129]:
print(accuracy_score(y_test,y_pred))
print(precision_score(y_test,y_pred,average='macro'))
print(recall_score(y_test,y_pred,average='macro'))

0.9411666666666667
0.9424641636310972
0.9409447483014526


In [130]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred,digits=4))


              precision    recall  f1-score   support

           0     0.9668    0.9831    0.9749       237
           1     0.8875    0.9261    0.9064       230
           2     0.9856    0.9276    0.9557       221
           3     0.9451    0.9256    0.9353       242
           4     0.9286    0.9043    0.9163       230
           5     0.9307    0.9267    0.9287       232
           6     0.9668    0.8793    0.9210       232
           7     0.8793    0.9273    0.9027       220
           8     0.8840    0.9779    0.9286       226
           9     0.9949    0.8750    0.9311       224
          10     0.9355    0.9144    0.9248       222
          11     0.9524    0.9649    0.9586       228
          12     0.9617    0.9496    0.9556       238
          13     0.9454    0.9574    0.9514       235
          14     0.9500    0.9248    0.9372       226
          15     0.9662    0.9502    0.9582       241
          16     0.9130    0.9830    0.9467       235
          17     0.8838    

In [131]:
print(confusion_matrix(y_test, y_pred))

[[233   1   0   0   0   0   0   0   0   0   0   1   1   0   0   0   0   1
    0   0   0   0   0   0   0   0]
 [  0 213   0   0   0   0   0   4   0   0   1   0   0   0   0   0   0   3
    6   0   0   3   0   0   0   0]
 [  0   0 205   0   4   0   1   1   3   0   0   1   0   0   2   0   0   1
    0   0   2   0   1   0   0   0]
 [  0   2   0 224   0   0   1   1   1   0   1   0   0   6   1   0   1   2
    0   0   2   0   0   0   0   0]
 [  0   5   0   0 208   5   1   0   0   0   0   4   1   0   0   0   1   0
    2   0   0   0   0   2   0   1]
 [  0   1   0   1   1 215   0   0   1   0   0   0   0   0   0   5   0   0
    2   2   0   0   0   1   3   0]
 [  0   1   0   0   3   0 204   2   0   0   1   3   1   1   4   0   7   2
    0   0   1   2   0   0   0   0]
 [  0   5   0   1   1   0   0 204   0   0   1   1   1   0   0   0   0   5
    0   0   0   0   1   0   0   0]
 [  0   0   0   0   0   1   0   0 221   0   0   0   0   0   0   1   1   0
    0   0   0   0   0   0   1   1]
 [  0   0   0   0  

# Manual Grid Search
import itertools
#### Function to create ANN model
def create_model(neurons1=128, neurons2=64, activation='relu', learning_rate=0.001, dropout_rate=0.1):

    model = Sequential()
    
    model.add(Dense(neurons1, activation=activation, input_shape=(X_train.shape[1],)))
    
    model.add(Dropout(dropout_rate))
    
    model.add(Dense(neurons2, activation=activation))
    
    model.add(Dropout(dropout_rate))
    
    model.add(Dense(len(y.iloc[:,0].unique()), activation='softmax'))  # softmax for multi-class classification
    
    optimizer = Adam(learning_rate=learning_rate)
    
    model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    
    return model

#### Hyperparameter options

neurons1_list = [64, 128]

neurons2_list = [32, 64]

activation_list = ['relu', 'tanh']

learning_rate_list = [0.001, 0.01]

dropout_list = [0.1, 0.2]

#### Store results
results = []

#### Manual grid search

for neurons1, neurons2, activation, lr, dropout_rate in itertools.product(
        neurons1_list, neurons2_list, activation_list, learning_rate_list, dropout_list
):
   
    print(f"Training with neurons1={neurons1}, neurons2={neurons2}, activation={activation}, lr={lr}, dropout={dropout_rate}")
    
    model = create_model(neurons1, neurons2, activation, lr, dropout_rate)
    
    history = model.fit(X_train, y_train, validation_split=0.2, epochs=50, batch_size=32, verbose=0)
    
    val_acc = max(history.history['val_accuracy'])
    
    results.append((neurons1, neurons2, activation, lr, dropout_rate, val_acc))
    
    print(f"Validation Accuracy: {val_acc:.4f}")

#### Find best hyperparameters
best_params = max(results, key=lambda x: x[5])

print("\nBest Hyperparameters:")

print(f"neurons1={best_params[0]}, neurons2={best_params[1]}, activation=
{best_params[2]}, learning_rate={best_params[3]}, dropout_rate={best_params[4]}")

print(f"Best Validation Accuracy: {best_params[5]:.4f}")

#### Train final model with best hyperparameters

final_model = create_model(best_params[0], best_params[1], best_params[2], 

best_params[3], best_params[4])

final_model.fit(X_train, y_train, epochs=50, batch_size=32, validation_split=0.2, verbose=2)

#### Evaluate on test set
loss, accuracy = final_model.evaluate(X_test, y_test)

print(f"Test Accuracy: {accuracy:.4f}")


In [49]:
!pip install scikeras

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.0.1 -> 25.2
[notice] To update, run: C:\Program Files\Python313\python.exe -m pip install --upgrade pip


In [55]:

# Function to create ANN model
def create_model(neurons1=128, neurons2=64, activation='relu', learning_rate=0.001, dropout_rate=0.1):
    model = Sequential()
    model.add(Dense(neurons1, activation=activation, input_shape=(X_train.shape[1],)))
    model.add(Dropout(dropout_rate))
    model.add(Dense(neurons2, activation=activation))
    model.add(Dropout(dropout_rate))
    model.add(Dense(len(np.unique(y_train)), activation='softmax'))
    optimizer = Adam(learning_rate=learning_rate)
    model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    return model

# Wrap model with KerasClassifier
clf = KerasClassifier(model=create_model, verbose=0, epochs=50, batch_size=32)

# Define hyperparameter grid
param_grid = {
    "model__neurons1": [64, 128],
    "model__neurons2": [32, 64],
    "model__activation": ['relu', 'tanh'],
    "model__learning_rate": [0.001, 0.01],
    "model__dropout_rate": [0.1, 0.2]
}
# Grid search
grid = GridSearchCV(estimator=clf, param_grid=param_grid, cv=3, n_jobs=-1, scoring="accuracy")
grid_result = grid.fit(X_train, y_train)

# Print best params
print("Best Hyperparameters:", grid_result.best_params_)
print("Best Cross-Validation Accuracy: {:.4f}".format(grid_result.best_score_))

# Evaluate on test set
best_model = grid_result.best_estimator_
test_accuracy = best_model.score(X_test, y_test)
print("Test Accuracy: {:.4f}".format(test_accuracy))

Best Hyperparameters: {'model__activation': 'relu', 'model__dropout_rate': 0.1, 'model__learning_rate': 0.001, 'model__neurons1': 128, 'model__neurons2': 64}
Best Cross-Validation Accuracy: 0.9476
Test Accuracy: 0.9575


In [132]:

#  Build a best ANN Model
Tuned_model = Sequential([
    Dense(128, activation='relu', input_shape=(X.shape[1],)),
    Dropout(0.1),
    Dense(64, activation='relu'),
    Dropout(0.1),
    Dense(len(y['letter'].unique()), activation='softmax')
])
optimizer = Adam(learning_rate=0.001)
Tuned_model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
# adam learning rate is 0.001 default
history = Tuned_model.fit(X_train, y_train, validation_split=0.2, epochs=30, batch_size=32)

Epoch 1/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.4848 - loss: 1.8140 - val_accuracy: 0.7150 - val_loss: 1.0265
Epoch 2/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.7024 - loss: 0.9851 - val_accuracy: 0.7889 - val_loss: 0.7494
Epoch 3/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.7577 - loss: 0.7858 - val_accuracy: 0.8164 - val_loss: 0.6197
Epoch 4/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.7845 - loss: 0.6914 - val_accuracy: 0.8396 - val_loss: 0.5346
Epoch 5/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.8116 - loss: 0.6121 - val_accuracy: 0.8532 - val_loss: 0.4831
Epoch 6/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8309 - loss: 0.5508 - val_accuracy: 0.8743 - val_loss: 0.4279
Epoch 7/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8424 - loss: 0.5040 - val_accuracy: 0.8804 - val_loss: 0.3951
Epoch 8/30
350/350 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.8501 - loss: 0.4738 - val_accuracy: 0.

In [133]:
y_pred_tune=Tuned_model.predict(X_test)
y_pred_tune = np.argmax(y_pred_tune, axis=1)

188/188 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step


In [134]:
y_pred_tune

array([ 8,  2, 10, ..., 20,  1, 20], shape=(6000,))

In [135]:
print(accuracy_score(y_test,y_pred_tune))
print(precision_score(y_test,y_pred_tune,average='macro'))
print(recall_score(y_test,y_pred_tune,average='macro'))

0.9453333333333334
0.9458561736561522
0.9452830679205502


From the hyper pearameter tuning we get the best values as

Best Hyperparameters:

neurons1=128, neurons2=64, activation=relu, learning_rate=0.001, dropout_rate=0.1
Best Validation Accuracy: 0.9453

* The default ANN achieved ~94.1% validation accuracy, showing strong baseline performance.
* After hyperparameter tuning with GridSearchCV, the best model achieved ~94.6% validation accuracy.
* The tuned model used 128–64 neurons with relu activation, dropout of 0.1, and a learning rate of 0.001.
* Compared to the baseline, tuning improved validation accuracy by ~0.5% and reduced validation loss, showing better generalization.
* This demonstrates that hyperparameter tuning can significantly improve model performance by optimizing the network’s structure and learning behavior.